File: python-elasticemail-examples/examples/email_status.py
"""
Email Status Example

Both ids are returned by every send call.

Usage: python examples/email_status.py <transactionId> [messageId]
"""

import os
import sys

import ElasticEmail

sys.path.insert(0, os.path.dirname(__file__))
from ee import get_configuration, print_api_error


def main():
    args = sys.argv[1:]
    if not args:
        print("Usage: python examples/email_status.py <transactionId> [messageId]", file=sys.stderr)
        sys.exit(1)

    transaction_id = args[0]
    message_id = args[1] if len(args) > 1 else None

    configuration = get_configuration()

    with ElasticEmail.ApiClient(configuration) as api_client:
        emails_api = ElasticEmail.EmailsApi(api_client)

        try:
            status = emails_api.emails_by_transactionid_status_get(
                transaction_id,
                show_failed=True,
                show_sent=True,
                show_delivered=True,
                show_pending=True,
                show_opened=True,
                show_clicked=True,
            )
            print("=== Transaction status ===")
            print("Status:     ", status.status)
            print("Recipients: ", status.recipients_count)
            print("Sent:       ", status.sent_count, status.sent or [])
            print("Delivered:  ", status.delivered_count, status.delivered or [])
            print("Pending:    ", status.pending_count)
            print("Opened:     ", status.opened_count)
            print("Clicked:    ", status.clicked_count)
            print("Failed:     ", status.failed_count, status.failed or [])
        except ElasticEmail.ApiException as e:
            print_api_error("fetch status", e)
            sys.exit(1)

        if message_id:
            try:
                data = emails_api.emails_by_msgid_view_get(message_id)
                preview = data.preview
                message_status = data.status
                print("\n=== Message ===")
                print("From:    ", preview.var_from if preview else None)
                print("Subject: ", preview.subject if preview else None)
                if message_status:
                    print("Status:  ", message_status.status_name, message_status.date_sent or "")
                body = (preview.body if preview else None) or ""
                print("Body preview:", body[:200] + "..." if len(body) > 200 else body)
            except ElasticEmail.ApiException as e:
                print_api_error("fetch message", e)


if __name__ == "__main__":
    main()

File: dotnet-elasticemail-examples/Examples/EmailStatus.cs
namespace ElasticEmailExamples;

using ElasticEmail.Api;
using ElasticEmail.Client;

public static class EmailStatus
{
    // Usage: dotnet run -- email-status <transactionId> [messageId]
    // Both ids are returned by every send call.
    public static async Task RunAsync(string[] args)
    {
        if (args.Length == 0)
        {
            Console.Error.WriteLine("Usage: dotnet run -- email-status <transactionId> [messageId]");
            Environment.Exit(1);
        }

        var transactionId = args[0];
        var messageId = args.Length > 1 ? args[1] : null;
        var emailsApi = new EmailsApi(Ee.Config());

        try
        {
            var data = await emailsApi.EmailsByTransactionidStatusGetAsync(
                transactionId,
                showFailed: true,
                showSent: true,
                showDelivered: true,
                showPending: true,
                showOpened: true,
                showClicked: true);

            Console.WriteLine("=== Transaction status ===");
            Console.WriteLine($"Status:      {data.Status}");
            Console.WriteLine($"Recipients:  {data.RecipientsCount}");
            Console.WriteLine($"Sent:        {data.SentCount} [{Join(data.Sent)}]");
            Console.WriteLine($"Delivered:   {data.DeliveredCount} [{Join(data.Delivered)}]");
            Console.WriteLine($"Pending:     {data.PendingCount}");
            Console.WriteLine($"Opened:      {data.OpenedCount}");
            Console.WriteLine($"Clicked:     {data.ClickedCount}");
            Console.WriteLine($"Failed:      {data.FailedCount} [{Join(data.Failed?.Select(f => f.Address))}]");
        }
        catch (ApiException e)
        {
            Ee.Fail("fetch status", e);
        }

        if (messageId == null)
        {
            return;
        }

        try
        {
            var data = await emailsApi.EmailsByMsgidViewGetAsync(messageId);
            Console.WriteLine("\n=== Message ===");
            Console.WriteLine($"From:     {data.Preview?.From}");
            Console.WriteLine($"Subject:  {data.Preview?.Subject}");
            Console.WriteLine($"Status:   {data.Status?.StatusName} {data.Status?.DateSent:O}");
            var body = data.Preview?.Body ?? "";
            Console.WriteLine($"Body preview: {(body.Length > 200 ? body[..200] + "..." : body)}");
        }
        catch (ApiException e)
        {
            Ee.PrintApiError("fetch message", e);
        }
    }

    private static string Join(IEnumerable<string>? values) => string.Join(", ", values ?? Enumerable.Empty<string>());
}
